docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
 
 ### 1.7 Optionality
 
-Symbol `?` can mark a type as optional.  
-Optionality means the value behind type can be either present or nil.
+Nullity (`null`, `undefined`, `nil`) is a primary source of bugs in software.
+Being able to guarantee that a value will never be null makes it easier to write safe
+code without constantly having to take nullity into account.
 
-Rules of optionality apply to the entire new container type of `type?` and not
-the value behind it (`type`).  
+We are going to try and design Wing without allowing users to explicitly assign or use the `nil`
+value. Instead, the language offers syntax to support the various use cases that arise when dealing
+with *optionality*.
 
-Using the `??` operator allows one to safely access the value behind an optional
-variable by always providing a default, in case the variable is nil. This forces
-a value to be present for the l-value (left hand side of the assignment operator
-and guarantees what's returned is type-stripped from its `?` keyword).
+Here's a quick summary of how optionality works in Wing:
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+* `x: T?` marks `x` as "optional of T". This means that `x` can store a value of type `T` but it can also be in a state where there is no value stored there.
+* To test if `x` has a value or not, use `x?`, which returns a `bool`. If x has a value, `x?` will be `true`, otherwise it will be `false`.
+* The `if let y = x { } else { }` is a special if statement that binds `y` inside the block only if `x` includes a value. Otherwise, the `else` block will be executed.
+* The `x?.y` notation can be used to access object fields only if there is a value. It returns an undefined optional if the value is not defined.
+* The `x ?? y` notation will return the value in `x` (if there is one) or `y` otherwise (`y` must be of type `T`, not `T?`).
+* A default value can also be supplied using `= y` in the declaration of the field or argument, which implies that the type will always be defined.
```

Comment:
This introduction is awesome (I appreciate the bulleted summary) - though I think it's missing one critical piece, which is an example that demonstrates why nullity is modeled in the language in the first place. Namely, by shifting null-pointer-exceptions from being a runtime error to a compile time error instead:

```
let increment = (x: num, by: num?): num => {
  return x + by; // Error: "by" is `num?` (optional type), it must be unwrapped first
};
```